File: rdna3_kawpow/dagcache.py
"""On-disk caching of the per-epoch light cache and full DAG.

A fresh start spends ~40-80 s building the epoch's light cache (pure-Python
``mkcache``) and then generating the multi-GB DAG on the GPU. Both are a pure
function of ``(epoch, seed, parents)``, so once produced they can be persisted
and reloaded on the next start, making restarts within an epoch effectively
instant.

Two artifacts are cached side by side, keyed by ``(variant, epoch, parents)``:

* the **light cache** (tens of MiB) -- needed both to regenerate the DAG and for
  host-side solution re-checking, and
* the **full DAG** (multi-GB) -- so the GPU generation pass can be skipped too.

Each file carries a small validated header (magic, format version, epoch,
parents, payload length and the 32-byte seed). Any mismatch or short/corrupt
file is silently treated as a miss and regenerated, so a stale cache can never
feed the miner the wrong data. Writes go to a ``.tmp`` sibling and are atomically
renamed, and old epochs are pruned so the directory does not grow without bound.

All disk access here is best-effort: callers wrap save/load so a full disk or a
permission error degrades to "regenerate", never a crash.
"""

import glob
import os
import struct

MAGIC_LIGHT = b"RKLC"        # rdna3-kawpow light cache
MAGIC_DAG = b"RKDG"          # rdna3-kawpow full DAG
VERSION = 1                  # bump to invalidate every existing cache file

# magic, format version, epoch, parents, payload length, 32-byte seed
_HEADER_FMT = "<4sIIIQ32s"
HEADER_SIZE = struct.calcsize(_HEADER_FMT)

# Retention: epochs advance rarely (every 7500 blocks), so a couple of DAGs and a
# few light caches is plenty to cover an epoch flip without unbounded growth.
KEEP_DAG = 2
KEEP_LIGHT = 4


def default_dir():
    """Persistent cache location (NOT the system tempdir, which may be wiped)."""
    base = os.environ.get("RDNA3_KAWPOW_CACHE_DIR")
    if base:
        return base
    root = os.environ.get("LOCALAPPDATA") or os.path.join(
        os.path.expanduser("~"), ".cache")
    return os.path.join(root, "rdna3_kawpow", "dagcache")


def _seed32(seed):
    return (bytes(seed or b"") + b"\x00" * 32)[:32]


def _pack_header(magic, epoch, parents, payload_len, seed):
    return struct.pack(_HEADER_FMT, magic, VERSION, epoch, parents,
                       payload_len, _seed32(seed))


def _check_header(raw, magic, epoch, parents, seed):
    """Return the payload length if the header matches expectations, else None."""
    if len(raw) != HEADER_SIZE:
        return None
    m, ver, ep, par, payload_len, sd = struct.unpack(_HEADER_FMT, raw)
    if (m != magic or ver != VERSION or ep != epoch or par != parents
            or sd != _seed32(seed)):
        return None
    return payload_len


class _Writer:
    """Write to a ``.tmp`` sibling and atomically rename on commit()."""

    def __init__(self, final_path):
        self.final = final_path
        self.tmp = final_path + ".tmp"
        self.f = open(self.tmp, "wb")

    def write(self, data):
        self.f.write(data)

    def commit(self):
        self.f.close()
        os.replace(self.tmp, self.final)

    def abort(self):
        try:
            self.f.close()
        finally:
            try:
                os.remove(self.tmp)
            except OSError:
                pass


class Cache:
    """Disk cache rooted at a directory; one instance per miner."""

    def __init__(self, directory=None):
        self.dir = directory or default_dir()
        os.makedirs(self.dir, exist_ok=True)

    # --- paths ---
    def _path(self, prefix, variant, epoch, parents):
        return os.path.join(self.dir, f"{prefix}-{variant}-e{epoch}-p{parents}.bin")

    def _prune(self, prefix, variant, keep):
        # Best-effort: a file vanishing mid-scan must never mask a successful save.
        try:
            pattern = os.path.join(self.dir, f"{prefix}-{variant}-e*.bin")
            stamped = []
            for f in glob.glob(pattern):
                try:
                    stamped.append((os.path.getmtime(f), f))
                except OSError:
                    pass
            stamped.sort(reverse=True)
            for _, old in stamped[keep:]:
                try:
                    os.remove(old)
                except OSError:
                    pass
        except OSError:
            pass

    # --- light cache (small; read/written whole) ---
    def load_light(self, variant, epoch, parents, seed):
        """Return the cached light-cache bytes, or None on miss/mismatch."""
        try:
            with open(self._path("light", variant, epoch, parents), "rb") as f:
                payload_len = _check_header(f.read(HEADER_SIZE), MAGIC_LIGHT,
                                            epoch, parents, seed)
                if payload_len is None:
                    return None
                data = f.read(payload_len)
                return data if len(data) == payload_len else None
        except OSError:
            return None

    def save_light(self, variant, epoch, parents, seed, data):
        w = _Writer(self._path("light", variant, epoch, parents))
        try:
            w.write(_pack_header(MAGIC_LIGHT, epoch, parents, len(data), seed))
            w.write(data)
            w.commit()
        except BaseException:
            w.abort()
            raise
        self._prune("light", variant, KEEP_LIGHT)

    # --- full DAG (multi-GB; streamed in chunks by the caller) ---
    def has_dag(self, variant, epoch, parents, seed, full_size):
        """True iff a DAG file exists whose header matches and payload == full_size."""
        try:
            with open(self._path("dag", variant, epoch, parents), "rb") as f:
                payload_len = _check_header(f.read(HEADER_SIZE), MAGIC_DAG,
                                            epoch, parents, seed)
            return payload_len == full_size
        except OSError:
            return False

    def open_dag_read(self, variant, epoch, parents):
        """Open the DAG file positioned at the payload start (call has_dag first)."""
        f = open(self._path("dag", variant, epoch, parents), "rb")
        f.seek(HEADER_SIZE)
        return f

    def open_dag_write(self, variant, epoch, parents, seed, full_size):
        """Return a _Writer with the DAG header already written; caller streams payload."""
        w = _Writer(self._path("dag", variant, epoch, parents))
        try:
            w.write(_pack_header(MAGIC_DAG, epoch, parents, full_size, seed))
        except BaseException:
            w.abort()
            raise
        return w

    def prune_dag(self, variant):
        self._prune("dag", variant, KEEP_DAG)

File: rust/src/dagcache.rs
//! On-disk caching of the per-epoch light cache and full DAG (port of
//! `rdna3_kawpow/dagcache.py`).
//!
//! A fresh start spends ~40-80 s building the light cache + generating the multi-GB
//! DAG. Both are a pure function of `(variant, epoch, parents, seed)`, so they are
//! persisted and reloaded to make restarts within an epoch near-instant.
//!
//! Each file carries a small validated header (magic, version, epoch, parents,
//! payload length, 32-byte seed). Any mismatch or short/corrupt file is treated as
//! a miss and regenerated, so a stale cache can never feed the miner wrong data.
//! Writes go to a `.tmp` sibling and are atomically renamed; old epochs are pruned.

use std::fs::{self, File};
use std::io::{self, BufWriter, Read, Seek, SeekFrom, Write};
use std::path::{Path, PathBuf};

const MAGIC_LIGHT: &[u8; 4] = b"RKLC";
const MAGIC_DAG: &[u8; 4] = b"RKDG";
const VERSION: u32 = 1; // bump to invalidate every existing cache file

// magic(4) | version(4) | epoch(4) | parents(4) | payload_len(8) | seed(32)
pub const HEADER_SIZE: usize = 4 + 4 + 4 + 4 + 8 + 32;

pub const KEEP_DAG: usize = 2;
pub const KEEP_LIGHT: usize = 4;

/// Persistent cache location (NOT the system tempdir, which may be wiped).
pub fn default_dir() -> PathBuf {
    if let Ok(base) = std::env::var("RDNA3_KAWPOW_CACHE_DIR") {
        return PathBuf::from(base);
    }
    let root = std::env::var("LOCALAPPDATA")
        .map(PathBuf::from)
        .unwrap_or_else(|_| {
            let home = std::env::var("HOME").unwrap_or_else(|_| ".".into());
            PathBuf::from(home).join(".cache")
        });
    root.join("rdna3_kawpow").join("dagcache")
}

fn seed32(seed: &[u8]) -> [u8; 32] {
    let mut s = [0u8; 32];
    let n = seed.len().min(32);
    s[..n].copy_from_slice(&seed[..n]);
    s
}

fn pack_header(magic: &[u8; 4], epoch: u32, parents: u32, payload_len: u64, seed: &[u8]) -> [u8; HEADER_SIZE] {
    let mut h = [0u8; HEADER_SIZE];
    h[0..4].copy_from_slice(magic);
    h[4..8].copy_from_slice(&VERSION.to_le_bytes());
    h[8..12].copy_from_slice(&epoch.to_le_bytes());
    h[12..16].copy_from_slice(&parents.to_le_bytes());
    h[16..24].copy_from_slice(&payload_len.to_le_bytes());
    h[24..56].copy_from_slice(&seed32(seed));
    h
}

/// Return the payload length iff the header matches expectations.
fn check_header(raw: &[u8], magic: &[u8; 4], epoch: u32, parents: u32, seed: &[u8]) -> Option<u64> {
    if raw.len() != HEADER_SIZE {
        return None;
    }
    if &raw[0..4] != magic {
        return None;
    }
    let ver = u32::from_le_bytes(raw[4..8].try_into().unwrap());
    let ep = u32::from_le_bytes(raw[8..12].try_into().unwrap());
    let par = u32::from_le_bytes(raw[12..16].try_into().unwrap());
    let payload_len = u64::from_le_bytes(raw[16..24].try_into().unwrap());
    let sd = &raw[24..56];
    if ver != VERSION || ep != epoch || par != parents || sd != seed32(seed) {
        return None;
    }
    Some(payload_len)
}

/// Write to a `.tmp` sibling and atomically rename on `commit()`.
pub struct Writer {
    final_path: PathBuf,
    tmp: PathBuf,
    f: Option<BufWriter<File>>,
}

impl Writer {
    fn create(final_path: PathBuf) -> io::Result<Self> {
        let tmp = final_path.with_extension("bin.tmp");
        let f = BufWriter::new(File::create(&tmp)?);
        Ok(Writer {
            final_path,
            tmp,
            f: Some(f),
        })
    }

    pub fn write_all(&mut self, data: &[u8]) -> io::Result<()> {
        self.f.as_mut().unwrap().write_all(data)
    }

    pub fn commit(mut self) -> io::Result<()> {
        let mut f = self.f.take().unwrap();
        f.flush()?;
        drop(f);
        fs::rename(&self.tmp, &self.final_path)
    }

    pub fn abort(mut self) {
        self.f.take();
        let _ = fs::remove_file(&self.tmp);
    }
}

/// Disk cache rooted at a directory; one instance per miner.
pub struct Cache {
    pub dir: PathBuf,
}

impl Cache {
    pub fn new(directory: Option<PathBuf>) -> io::Result<Cache> {
        let dir = directory.unwrap_or_else(default_dir);
        fs::create_dir_all(&dir)?;
        Ok(Cache { dir })
    }

    fn path(&self, prefix: &str, variant: &str, epoch: u32, parents: u32) -> PathBuf {
        self.dir
            .join(format!("{prefix}-{variant}-e{epoch}-p{parents}.bin"))
    }

    fn prune(&self, prefix: &str, variant: &str, keep: usize) {
        let needle = format!("{prefix}-{variant}-e");
        let mut stamped: Vec<(std::time::SystemTime, PathBuf)> = Vec::new();
        let Ok(rd) = fs::read_dir(&self.dir) else {
            return;
        };
        for entry in rd.flatten() {
            let name = entry.file_name();
            let name = name.to_string_lossy();
            if name.starts_with(&needle) && name.ends_with(".bin") {
                if let Ok(meta) = entry.metadata() {
                    if let Ok(mtime) = meta.modified() {
                        stamped.push((mtime, entry.path()));
                    }
                }
            }
        }
        stamped.sort_by(|a, b| b.0.cmp(&a.0)); // newest first
        for (_, old) in stamped.into_iter().skip(keep) {
            let _ = fs::remove_file(old);
        }
    }

    // --- light cache (small; read/written whole) ---
    pub fn load_light(&self, variant: &str, epoch: u32, parents: u32, seed: &[u8]) -> Option<Vec<u8>> {
        let mut f = File::open(self.path("light", variant, epoch, parents)).ok()?;
        let mut header = [0u8; HEADER_SIZE];
        f.read_exact(&mut header).ok()?;
        let payload_len = check_header(&header, MAGIC_LIGHT, epoch, parents, seed)? as usize;
        let mut data = vec![0u8; payload_len];
        f.read_exact(&mut data).ok()?;
        Some(data)
    }

    pub fn save_light(&self, variant: &str, epoch: u32, parents: u32, seed: &[u8], data: &[u8]) -> io::Result<()> {
        let mut w = Writer::create(self.path("light", variant, epoch, parents))?;
        let res = (|| -> io::Result<()> {
            w.write_all(&pack_header(MAGIC_LIGHT, epoch, parents, data.len() as u64, seed))?;
            w.write_all(data)
        })();
        match res {
            Ok(()) => w.commit()?,
            Err(e) => {
                w.abort();
                return Err(e);
            }
        }
        self.prune("light", variant, KEEP_LIGHT);
        Ok(())
    }

    // --- full DAG (multi-GB; streamed in chunks by the caller) ---
    pub fn has_dag(&self, variant: &str, epoch: u32, parents: u32, seed: &[u8], full_size: u64) -> bool {
        let Ok(mut f) = File::open(self.path("dag", variant, epoch, parents)) else {
            return false;
        };
        let mut header = [0u8; HEADER_SIZE];
        if f.read_exact(&mut header).is_err() {
            return false;
        }
        check_header(&header, MAGIC_DAG, epoch, parents, seed) == Some(full_size)
    }

    /// Open the DAG file positioned at the payload start (call `has_dag` first).
    pub fn open_dag_read(&self, variant: &str, epoch: u32, parents: u32) -> io::Result<File> {
        let mut f = File::open(self.path("dag", variant, epoch, parents))?;
        f.seek(SeekFrom::Start(HEADER_SIZE as u64))?;
        Ok(f)
    }

    /// A `Writer` with the DAG header already written; caller streams the payload.
    pub fn open_dag_write(&self, variant: &str, epoch: u32, parents: u32, seed: &[u8], full_size: u64) -> io::Result<Writer> {
        let mut w = Writer::create(self.path("dag", variant, epoch, parents))?;
        if let Err(e) = w.write_all(&pack_header(MAGIC_DAG, epoch, parents, full_size, seed)) {
            w.abort();
            return Err(e);
        }
        Ok(w)
    }

    pub fn prune_dag(&self, variant: &str) {
        self.prune("dag", variant, KEEP_DAG);
    }
}

/// Convenience: does `path` look like a cache file we manage?
pub fn is_cache_file(path: &Path) -> bool {
    path.extension().map(|e| e == "bin").unwrap_or(false)
}
